src/connectivity/percolation.c

```diff
@@ -0,0 +1,281 @@
+    igraph_integer_t parent, child;
+    if (VECTOR(*sizes)[a] < VECTOR(*sizes)[b]) {
+        parent = b;
+        child = a;
+    } else {
+        parent = a;
+        child = b;
+    }
+
+    // Make a child of b
+    VECTOR(*links)[child] = parent;
+    VECTOR(*sizes)[parent] += VECTOR(*sizes)[child];
+
+    // If made new biggest component, update biggest
+    if (VECTOR(*sizes)[parent] >= *biggest) {
+        *biggest = VECTOR(*sizes)[parent];
+    }
+}
+
+/**
+ * \function igraph_i_edge_list_percolation
+ * \brief Gives the size of the largest connected component as edges are added.
+ * \param edges Vector of edges, will be added in order.
+ * \param output output[i] is the size of the largest connected component after edges[i] is added
+ */
+
+static igraph_error_t edge_list_percolation(
+        const igraph_vector_int_t *edges,
+        igraph_vector_int_t* output) {
+
+    igraph_integer_t biggest = 0;
+    igraph_integer_t lower, upper;
+
+    igraph_vector_int_minmax(edges, &lower, &upper);
+
+    if (lower < 0) {
+        IGRAPH_ERROR("Negative number in edge list.", IGRAPH_EINVVID);
+    }
+
+    igraph_vector_int_t sizes;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&sizes, upper + 1);
+
+    igraph_vector_int_t links;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&links, upper + 1);
+
+    for (igraph_integer_t i = 0; i < upper + 1; i++) {
+        VECTOR(sizes)[i] = 1;
+        VECTOR(links)[i] = i;
+    }
+
+    igraph_integer_t edge_count = igraph_vector_int_size(edges);
+    if (edge_count % 2 == 1) {
+        IGRAPH_ERROR("Invalid edge list, odd number of elements.", IGRAPH_EINVAL);
+    }
+    edge_count >>= 1;
+    IGRAPH_CHECK(igraph_vector_int_resize(output, edge_count));
+
+    for (igraph_integer_t i = 0; i < edge_count; i++) {
+        percolate_edge(&links, &sizes, &biggest, VECTOR(*edges)[2 * i], VECTOR(*edges)[2 * i + 1]);
+        VECTOR(*output)[i] = biggest;
+    }
+    igraph_vector_int_destroy(&sizes);
+    igraph_vector_int_destroy(&links);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_bond_percolation
+ * \brief calculates the bond-percolation curve of a graph.
+ * \param graph The input graph
+ * \param output output[i] is the size of the largest component after adding i+1 edges, will be created.
+ */
+
+igraph_error_t igraph_bond_percolation(
+        const igraph_t *graph,
+        igraph_vector_int_t * output,
+        igraph_vector_int_t * edge_indices) {
+
+    igraph_vector_int_t* internal_edge_indices = edge_indices;
+    igraph_vector_int_t new_edges;
+    igraph_bool_t generated_edges = false;
+
+    if (edge_indices == NULL) {
+        // Use random edge order
+        generated_edges = true;
+        igraph_integer_t size = igraph_ecount(graph);
+
+        IGRAPH_CHECK(igraph_vector_int_init_range(&new_edges, 0, size));
+
+        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_edges));
+
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_edges);
```

Comment:
```suggestion
        IGRAPH_CHECK(igraph_vector_int_init_range(&new_edges, 0, size));
        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_edges);
        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_edges));
```

There must be no possibility of an error being thrown between the `init` call and the corresponding `FINALLY` call.

Theoretically, if `igraph_vector_int_shuffle()` fails, `new_edges` won't be freed as it wasn't placed on the "finally" stack yet.

I know that in this case, `igraph_vector_int_shuffle()` can't _really_ fail. In fact, in the next major version the return type of this function will be changed to `void`: https://github.com/igraph/igraph/blob/4d325eeb39558583748b1d5d378ba5e11c02f156/CHANGELOG.md?plain=1#L13  Nevertheless, we should always keep the `init` and the `FINALLY` calls together, so that when we read the code later, it doesn't look like something's wrong.

Please fix this (you can simply commit my suggestion using the button above).